bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -151,25 +154,30 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 		return ctrl.Result{}, err
 	}
 
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(config, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
 	switch {
-	// Wait patiently for the infrastructure to be ready
+	// Wait for the infrastructure to be ready.
 	case !cluster.Status.InfrastructureReady:
-		log.Info("Infrastructure is not ready, waiting until ready.")
+		log.Info("Cluster infrastructure is not ready, waiting")
 		return ctrl.Result{}, nil
-	// bail super early if it's already ready
+	// Migrate plaintext data to secret.
+	case config.Status.BootstrapData != nil:
+		if err := r.storeBootstrapData(ctx, config, config.Status.BootstrapData); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, patchHelper.Patch(ctx, config)
```

Comment:
In migration case, this condition will always be true, so will switch always stops here and won't progress to check later conditions?

Reply:
Does it need to? If we have bootstrap data inline as a string, we need to convert it to a secret and stop doing anything else. Right?

Reply:
I added a check up there to only migrate if the secret isn't populate, shouldn't that fix this issue?

Reply:
Yeah thats true, the secretName == nil check will fix it. I missed that part.